### ISSUE TITLE
observability: make logs structural

### ISSUE THREAD

**Issue by BugenZhao** [risingwavelabs/risingwave#10426]
Currently when printing logs with `tracing`, we tend to `format` the runtime information into the message like this:

https://github.com/risingwavelabs/risingwave/blob/4f3bce2d024f958d1eeebc49a5556a81372fe824/src/meta/src/rpc/election_client.rs#L107-L122

Although intuitive, this approach lacks observability friendliness. It is hard to retrieve the `lease_id` from these messages after they're formatted, which makes it also challenging for us to build some analytics or alerting tools by consolidating related logs, traces, and metrics. 

Instead, we should prefer structural logs instead. This is enabled by "[recording fields](https://docs.rs/tracing/latest/tracing/#recording-fields)" with `tracing`. It looks like...

https://github.com/risingwavelabs/risingwave/blob/4f3bce2d024f958d1eeebc49a5556a81372fe824/src/utils/pgwire/src/pg_protocol.rs#L345-L352

By explicitly specifying fields, the tracing subscriber will be able to decide how to handle them.
- For local debugging, we can get them printed after the message just like they're formatted into the message. This is what we've already done.
- For cloud deployment, we can format each event into a structural format (like JSON, which is a common practice of Golang), so the logging services can collect them back in a structural format losslessly. 
Currently, we use Promtail to extract the "flatten" lines from the log file to Grafana Loki on the cloud, where structural fields cannot be parsed accurately.
- For tracing, the OTLP-compatible collector will also recognize these fields and enable us to query logs or spans with these fields. (#10000)

cc @fuyufjh @arkbriar @Nebulazhang

**Comment by BugenZhao** [comment]
> So this means we need to refactor all logs?

I think so, at least for those useful for debugging.

**Comment by fuyufjh** [comment]
+1

Using different log config for multiple deploying env is quite common.

> So this means we need to refactor all logs?

From my experience, it's almost impossible to enforce *every* log to be structured. However, we can enforce it for the most critical logs and newly-added ones. It still helps.

**Comment by arkbriar** [comment]
> From my experience, it's almost impossible to enforce _every_ log to be structured. However, we can enforce it for the most critical logs and newly-added ones. It still helps.

Agree. It's more feasible to start on critical ones and then refactor the others gradually. 

Besides, structured logs make everything based on logs easier, such as analysis on SQL logs.

**Comment by arkbriar** [comment]
> > From my experience, it's almost impossible to enforce _every_ log to be structured. However, we can enforce it for the most critical logs and newly-added ones. It still helps.
> 
> Agree. It's more feasible to start on critical ones and then refactor the others gradually.
> 
> Besides, structured logs make everything based on logs easier, such as analysis on SQL logs.

But I'm concern about if the log collector recognizes hybrid format since we won't unify all the logs at once. Do you have any idea? @Nebulazhang

**Comment by BugenZhao** [comment]
The format won't be "hybrid". They are always in JSON if the env var `RISINGWAVE_CLOUD` is set after #10633.

For example, if a log event is structural, we may get something like...

```rust
tracing::debug!(actor_id, ?state, "take snapshot")
```
```json
{
    "timestamp": "2023-06-28T17:17:35.404904+08:00",
    "level": "DEBUG",
    "fields": {
        "message": "take snapshot",
        "actor_id": 42, <- 👀
        "state": "[Nexmark(NexmarkSplit { split_index: 2, split_num: 4, start_offset: Some(315394) })]" <- 👀
    },
    "target": "risingwave_stream::executor::source::source_executor"
}
```

But if it's simply formatted into the message, we'll only get the `message` field.

```rust
tracing::debug!("take snapshot for actor {} with state: {:?}", actor_id, state)
```
```json
{
    "timestamp": "2023-06-28T17:17:35.404904+08:00",
    "level": "DEBUG",
    "fields": {
        "message": "take snapshot for actor 42 with state: [Nexmark(NexmarkSplit { split_index: 2, split_num: 4, start_offset: Some(315394) })]"
    },
    "target": "risingwave_stream::executor::source::source_executor"
}
```

Although it may be harder to analyze these unstructural logs, this format should still be considered consistent with the structural one.

**Comment by xxchan** [comment]
Are there any actionable sub-issues?

**Comment by BugenZhao** [comment]
> Are there any actionable sub-issues?

I have no idea currently. Maybe we need some practices of log analyzing in production to determine what needs prioritization.

**Comment by xxchan** [comment]
No actionable tasks now. BTW, we can add a section in the developer guide https://github.com/risingwavelabs/risingwave/issues/17436
